web3.js/src/connection.ts

```diff
@@ -2161,40 +2216,35 @@ export class Connection {
     simulatedSignatures: [],
   };
 
-  /** @internal */ _accountChangeSubscriptionCounter: number = 0;
-  /** @internal */ _accountChangeSubscriptions: {
-    [id: number]: AccountSubscriptionInfo;
-  } = {};
-
-  /** @internal */ _programAccountChangeSubscriptionCounter: number = 0;
-  /** @internal */ _programAccountChangeSubscriptions: {
-    [id: number]: ProgramAccountSubscriptionInfo;
-  } = {};
-
-  /** @internal */ _rootSubscriptionCounter: number = 0;
-  /** @internal */ _rootSubscriptions: {
-    [id: number]: RootSubscriptionInfo;
+  /** @internal */ _nextClientSubscriptionId: ClientSubscriptionId = 0;
+  /** @internal */ _subscriptionDisposeFunctionsByClientSubscriptionId: {
+    [clientSubscriptionId: ClientSubscriptionId]:
+      | SubscriptionDisposeFn
+      | undefined;
   } = {};
-
-  /** @internal */ _signatureSubscriptionCounter: number = 0;
-  /** @internal */ _signatureSubscriptions: {
-    [id: number]: SignatureSubscriptionInfo;
-  } = {};
-
-  /** @internal */ _slotSubscriptionCounter: number = 0;
-  /** @internal */ _slotSubscriptions: {
-    [id: number]: SlotSubscriptionInfo;
-  } = {};
-
-  /** @internal */ _logsSubscriptionCounter: number = 0;
-  /** @internal */ _logsSubscriptions: {
-    [id: number]: LogsSubscriptionInfo;
+  /** @internal */ _subscriptionCallbacksByServerSubscriptionId: {
+    [serverSubscriptionId: ServerSubscriptionId]:
+      | Set<SubscriptionConfig['callback']>
+      | undefined;
   } = {};
-
-  /** @internal */ _slotUpdateSubscriptionCounter: number = 0;
-  /** @internal */ _slotUpdateSubscriptions: {
-    [id: number]: SlotUpdateSubscriptionInfo;
+  /** @internal */ _subscriptionsByHash: {
+    [hash: SubscriptionConfigHash]: Subscription | undefined;
   } = {};
+  /**
+   * Special case.
+   * After a signature is processed, RPCs automatically dispose of the
+   * subscription on the server side. We need to track which of these
+   * subscriptions have been disposed in such a way, so that we know
+   * whether the client is dealing with a not-yet-processed signature
+   * (in which case we must tear down the server subscription) or an
+   * already-processed signature (in which case the client can simply
+   * clear out the subscription locally without telling the server).
+   *
+   * NOTE: There is a proposal to eliminate this special case, here:
+   * https://github.com/solana-labs/solana/issues/18892
+   */
+  /** @internal */ _subscriptionsAutoDisposedByRpc: Set<ServerSubscriptionId> =
```

Comment:
Welllll… before I do that, we need to adjudicate this: https://github.com/solana-labs/solana/issues/18892

I bolted this special case on deliberately so, as to make it easy to rip it out later, once we decide to go ahead addressing that issue.

Reply:
As for the flow, it's all crafted so that unsubscribes _do_ get sent, if:

* You call `remove` before getting any notification.
* You call `remove` when the only thing you've received so far is a `signatureReceived` notification

…but they _do not_ get sent, if:

* You call `remove` after having received a `signatureProcessed` or regular status notification.

This is all exercised in the tests. Look for ‘after an auto-disposing subscription has published a non-final notification.’

Reply:
Ah ok, I'm just happy to hear that it's setup to rip out later :) I'm fine with it is for now